core/src/main/java/com/scalar/db/transaction/consensuscommit/PreparedSnapshotHook.java

```diff
@@ -0,0 +1,10 @@
+package com.scalar.db.transaction.consensuscommit;
+
+public interface PreparedSnapshotHook {
```

Comment:
`PreparedSnapshotHook` sounds like a hook called after preparation to me. Actually, the state of the snapshot changes after preparation.
https://github.com/scalar-labs/scalardb/blob/037913934c5e7dca44262c280b48136300fdeb92/core/src/main/java/com/scalar/db/transaction/consensuscommit/Snapshot.java#L459

Maybe `BeforePreparationSnapshotHook` would be a better name?

In the future, we might add something like `BeforeValidationSnapshotHook` or `BeforeCommitSnapshotHook`.

What do you think?

Reply:
> PreparedSnapshotHook sounds like a hook called after preparation to me.

Indeed, I have the same concern slightly. https://github.com/scalar-labs/scalardb/pull/2271#discussion_r1792738799

It's possible that the prepared records are asynchronously written before a snapshot is sent. So, I was also wondering if `BeforePreparationSnapshotHook`-ish name was reasonable or not. But, I think it can mean the snapshot is sent before the prepare phase operation. So, I think it's okay. I'll change it to `BeforePreparationSnapshotHook`.

Reply:
Addressed it in https://github.com/scalar-labs/scalardb/pull/2271/commits/0fe41c9412a03a215c00f1f0bf30ae210493fd8f

Reply:
As Toshi said, snapshot states might be changed in the case of extra-write. So, we shouldn't trigger the hook asynchronously?

Reply:
@feeblefakie 

Ah, you're right... I just remembered that @brfrn169 and I discussed this case and planed to disable the snapshot hook feature when extra-write is enabled, since extra-write would be sunset-ed (so, this PR should've checked the configuration...).

Triggering the hook after preparation might increase the total latency. So, I think the invocation of the hook should be asynchronous without waiting for the completion of `prepare` phase operation, although it might increase the possibility of transactions that abort in `prepare` phase being handled by the snapshot handler to some extent. What do you think?

Reply:
One possible solution is for the Snapshot getters to return copies of the read-set and write-set to avoid the impact of state changes?

Reply:
@brfrn169 Oh, do you think the read-set or the write-set could be changed during the `prepare` or `validate` phase even if extra-write is out of the scope?

Reply:
For now, if extra-write is out of scope, the read-set and write-set should not change. So, if we don’t allow users to set `BeforePreparationSnapshotHook` when extra-write is enabled, that should be fine. However, I’m concerned that this might not be guaranteed in the future. Therefore, it might be safer for the getters to return copies of the read-set and write-set, since `BeforePreparationSnapshotHook` is executed asynchronously. Otherwise, thread-safety issues could arise. What do you think?

Reply:
@brfrn169 

> I’m concerned that this might not be guaranteed in the future.

I understand your concern. It seems reasonable.

Okay. Indeed, the micro-benchmark showed copying the map had some performance impact, but it didn't necessarily indicate ScalarDB application incurs significant performance impact. So, I'll copy the map in this PR just in case. And, I'll compare with a zero-copy version later in a ScalarDB application from the perspective of performance and resource consumption.

BTW, I assume we don't need to revert https://github.com/scalar-labs/scalardb/pull/2271/commits/71b40a4168c24dbfdbd8eb4ef7852b071cbc2395#diff-4e6eb5a222519a33681ef588e24c3a8eabc8bdec7e89e79bb5c96c3ce8898759R158-R176 since it's called synchronously. Is my understanding correct?

Reply:
@brfrn169 

> I’m concerned that this might not be guaranteed in the future.

Ah, in this case, we may need to pass changed read-set and write-set to the snapshot hook, although it depends. It would require CommitHandler to pass changed read-set and write-set to the snapshot hook after PREPARED records are written. So, I don't think just copying the maps is sufficient.

How about going with the current no-copy approach in this PR for now? If we need to change read-set and write-set after PREPARED records are written in the future, we can modify the implementation introduced by this PR to copy the maps and delay the timing to pass changed read-set and write
-set to the snapshot hook handler. What do you think?

Reply:
> I don’t think just copying the maps is sufficient.

Ah yes, if we assume the preparation, validation, and commit phases might change the state of the snapshot (read-set, write-set, etc.), I agree that simply copying the maps is not sufficient.

Ideally, to make the hook more general, I think we should pass an immutable state of the snapshot at the time suggested by the hook’s name.

For example:
- For `BeforePreparationSnapshotHook`, we should pass the immutable state before preparation.
- For `BeforeValidationSnapshotHook`, we should pass the immutable state before validation.
- For `BeforeCommitSnapshotHook`, we should pass the immutable state before commit.

(We won’t add `BeforeValidationSnapshotHook` and `BeforeCommitSnapshotHook` in this PR though.)

However, in the current implementation, the snapshot passed to the hook is mutable, and since the hook could be executed asynchronously, if the preparation, validation, and commit phases change the state of the snapshot, the hook could access the state at an unpredictable time.

> How about going with the current no-copy approach in this PR for now? If we need to change read-set and write-set after PREPARED records are written in the future, we can modify the implementation introduced by this PR to copy the maps and delay the timing to pass changed read-set and write
-set to the snapshot hook handler. What do you think?

I’m fine with that, but should we add some validation to check if extra-write is used when `BeforePreparationSnapshotHook` is set?

Reply:
> However, in the current implementation, the snapshot passed to the hook is mutable, and since the hook could be executed asynchronously, if the preparation, validation, and commit phases change the state of the snapshot, the hook could access the state at an unpredictable time.

I think you have a point. It's hard to create an immutable snapshot. So, do you think we should give up passing a snapshot and just passes the immutable read/write-sets like this?

```java
public class ReadWriteSets {
   // These are copied from Snapshot
   final Map<Key, TransactionResult> readSet;
   final Map<Key, Put> writeSet;
   final Map<Key, Delete> deleteSet;
}

public interface BeforePreparationReadWriteSetsHook {
  BeforePreparationReadWriteSetsHookFuture handle(ReadWriteSets readWriteSets);
}
```

I'm okay with it.

As for extra-write, yeah I'll add a validation for that in this PR later.

Reply:
> So, do you think we should give up passing a snapshot and just passes the immutable read/write-sets like this?

Yes, that’s what I had in mind. Thanks.

> As for extra-write, yeah I'll add a validation for that in this PR later.

Maybe, if we pass the immutable read/write-sets to the hook, we may be able to support extra-write as well.

The extra-write modifies the write-set in the preparation phase, but actually I don’t think we need to replicate it to the secondary site since it’s only for concurrency control.

Reply:
> The extra-write modifies the write-set in the preparation phase, but actually I don’t think we need to replicate it to the secondary site since it’s only for concurrency control.

It seems like this causes inconsistent `tx_id` between the primary site and the backup site. Records read by a transaction will be updated with the transaction ID in the primary site, while the same records won't be updated in the backup site.

> As for extra-write, yeah I'll add a validation for that in this PR later.

[Done](https://github.com/scalar-labs/scalardb/pull/2271/commits/5b771342cf4171716f95b2612ed3aef835634ffe)

Reply:
> > So, do you think we should give up passing a snapshot and just passes the immutable read/write-sets like this?
>
> Yes, that’s what I had in mind. Thanks.

[Done](https://github.com/scalar-labs/scalardb/pull/2271/commits/8f0817d88b87168095a743f5fa3f10d273aea655)

@brfrn169 I noticed `new ArrayList(map.entrySet())` is twice faster than returning `new HashMap(map)`. So, I added the method `List<Entry<Key, Put>> getKeysAndPutsInWriteSet()` instead of `Map<Key, Put> getWriteSetMap()`. I think the new method is very similar to the existing method `List<Put> getPutsInWriteSet()`. Probably we can replace the usage of `List<Put> getPutsInWriteSet()` with the new method. What do you think?

Reply:
>> As for extra-write, yeah I'll add a validation for that in this PR later.
>
> Done

On second thought, whether to prohibit EXTRA WRITE depends on the usage of write sets. On the above discussions, I was only thinking of the use case of replication. But, write set captured before `prepare` might work with EXTRA WRITE in other use cases. So,  I think a validation to prohibit using EXTRA WRITE should be handled by a plug-in side not in `scalardb:core`. Probably I should revert https://github.com/scalar-labs/scalardb/pull/2271/commits/8f0817d88b87168095a743f5fa3f10d273aea655 ...

Reply:
> > > So, do you think we should give up passing a snapshot and just passes the immutable read/write-sets like this?
> >
> > Yes, that’s what I had in mind. Thanks.

> [Done](https://github.com/scalar-labs/scalardb/pull/2271/commits/8f0817d88b87168095a743f5fa3f10d273aea655)

Oops. What I did was not along with this idea. I'll address this...

Reply:
> On second thought, whether to prohibit EXTRA WRITE depends on the usage of write sets. On the above discussions, I was only thinking of the use case of replication. But, write set captured before prepare might work with EXTRA WRITE in other use cases. So, I think a validation to prohibit using EXTRA WRITE should be handled by a plug-in side not in scalardb:core. Probably I should revert https://github.com/scalar-labs/scalardb/commit/8f0817d88b87168095a743f5fa3f10d273aea655 ...

[Reverted](https://github.com/scalar-labs/scalardb/pull/2271/commits/882e476250eb0917822e0f021104bdca11be87b3)